native/dsteqr.go

```diff
@@ -104,256 +99,266 @@ func (impl Implementation) Dsteqr(compz lapack.EigComp, n int, d, e, z []float64
+					}
+					continue
+				}
+
+				// If remaining matrix is 2×2, use Dlae2 to compute its eigensystem.
+				if m == l+1 {
+					if icompz > 0 {
+						d[l], d[l+1], work[l], work[n-1+l] = impl.Dlaev2(d[l], e[l], d[l+1])
+						impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward,
+							n, 2, work[l:], work[n-1+l:], z[l:], ldz)
+					} else {
+						d[l], d[l+1] = impl.Dlae2(d[l], e[l], d[l+1])
+					}
+					e[l] = 0
+					l += 2
+					if l > lend {
+						break
+					}
+					continue
+				}
 
-		// Form shift
-		g = (d[l+1] - p) / (2 * e[l])
-		r = impl.Dlapy2(g, 1)
-		g = d[m] - p + e[l]/(g+math.Copysign(r, g))
-		s = 1
-		c = 1
-		p = 0
+				if jtot == nmaxit {
+					break
+				}
+				jtot++
 
-		// Inner loop
-		for i := m - 1; i >= l; i-- {
-			f := s * e[i]
-			b := c * e[i]
-			c, s, r = impl.Dlartg(g, f)
-			if i != m-1 {
-				e[i+1] = r
-			}
-			g = d[i+1] - p
-			r = (d[i]-g)*s + 2*c*b
-			p = s * r
-			d[i+1] = g + p
-			g = c*r - b
+				// Form shift
+				g := (d[l+1] - p) / (2 * e[l])
+				r := impl.Dlapy2(g, 1)
+				g = d[m] - p + e[l]/(g+math.Copysign(r, g))
+				s := 1.0
+				c := 1.0
+				p = 0.0
 
-			// If eigenvectors are desired, then save rotations.
-			if icompz > 0 {
-				work[i] = c
-				work[n-1+i] = -s
-			}
-		}
-		// If eigenvectors are desired, then apply saved rotations.
-		if icompz > 0 {
-			mm := m - l + 1
-			impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward,
-				n, mm, work[l:], work[n-1+l:], z[l:], ldz)
-		}
-		d[l] -= p
-		e[l] = g
-		goto Fourty
+				// Inner loop
+				for i := m - 1; i >= l; i-- {
+					f := s * e[i]
+					b := c * e[i]
+					c, s, r = impl.Dlartg(g, f)
+					if i != m-1 {
+						e[i+1] = r
+					}
+					g = d[i+1] - p
+					r = (d[i]-g)*s + 2*c*b
+					p = s * r
+					d[i+1] = g + p
+					g = c*r - b
 
-		// Eigenvalue found.
-	Eighty:
-		d[l] = p
-		l++
-		if l <= lend {
-			goto Fourty
-		}
-		goto OneFourty
-	} else {
-		// QR Iteration.
-		// Look for small superdiagonal element.
-	Ninety:
-		if l != lend {
-			lendp1 := lend + 1
-			for m = l; m >= lendp1; m-- {
-				v := math.Abs(e[m-1])
-				if v*v <= (eps2*math.Abs(d[m])*math.Abs(d[m-1]) + safmin) {
-					goto OneTen
+					// If eigenvectors are desired, then save rotations.
+					if icompz > 0 {
+						work[i] = c
+						work[n-1+i] = -s
+					}
 				}
+				// If eigenvectors are desired, then apply saved rotations.
+				if icompz > 0 {
+					mm := m - l + 1
+					impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward,
+						n, mm, work[l:], work[n-1+l:], z[l:], ldz)
+				}
+				d[l] -= p
+				e[l] = g
 			}
-		}
-		m = lend
-	OneTen:
-		if m > lend {
-			e[m-1] = 0
-		}
-		p = d[l]
-		if m == l {
-			goto OneThirty
-		}
+		} else {
+			// QR Iteration.
+			// Look for small superdiagonal element.
+			for {
+				if l != lend {
+					for m = l; m > lend; m-- {
+						v := math.Abs(e[m-1])
+						if v*v <= (eps2*math.Abs(d[m])*math.Abs(d[m-1]) + safmin) {
+							break
+						}
+					}
+				} else {
+					m = lend
```

Comment:
Yes (just got to a machine). You accepted the equivalent C-ism in dlasq3.go which differed only that it was not wrapped by a condition - the else path is needed because in the original the goto is only executes in the affirmative path. Without the else path we get index out of range panics - experimental support for the theoretical truth.